blockchain/stake/treasury_test.go

```diff
@@ -0,0 +1,1740 @@
+		},
+		is:       IsTSpend,
+		expected: false,
+		check:    checkTSpend,
+	}, {
+		name: "tspend P2PKH",
+		createTx: func() *wire.MsgTx {
+			// OP_RETURN <32 byte random>
+			payload := make([]byte, chainhash.HashSize)
+			_, err := rand.Read(payload)
+			if err != nil {
+				panic(err)
+			}
+			builder := txscript.NewScriptBuilder()
+			builder.AddOp(txscript.OP_RETURN)
+			builder.AddData(payload)
+			opretScript, err := builder.Script()
+			if err != nil {
+				panic(err)
+			}
+			msgTx := wire.NewMsgTx()
+			msgTx.Version = wire.TxVersionTreasury
+			msgTx.AddTxOut(wire.NewTxOut(0, opretScript))
+
+			// OP_TGEN
+			privKey := secp256k1.NewPrivateKey(new(secp256k1.ModNScalar).SetInt(1))
+			pubKey := privKey.PubKey()
+			pkHash := dcrutil.Hash160(pubKey.SerializeCompressed())
+			p2pkhAddr, err := dcrutil.NewAddressPubKeyHash(pkHash,
+				chaincfg.MainNetParams(), dcrec.STEcdsaSecp256k1)
+
+			if err != nil {
+				panic(err)
+			}
+			p2shOpTrueScript, err := txscript.PayToAddrScript(p2pkhAddr)
+			if err != nil {
+				panic(err)
+			}
+			script := make([]byte, len(p2shOpTrueScript)+1)
+			script[0] = txscript.OP_TGEN
+			copy(script[1:], p2shOpTrueScript)
+			msgTx.AddTxOut(wire.NewTxOut(0, script))
+
+			// tspend
+			builder = txscript.NewScriptBuilder()
+			builder.AddData(validSignature)
+			builder.AddData(publicKey)
+			builder.AddOp(txscript.OP_TSPEND)
+			tspendScript, err := builder.Script()
+			if err != nil {
+				panic(err)
+			}
+
+			msgTx.AddTxIn(&wire.TxIn{
+				// Stakebase transactions have no
+				// inputs, so previous outpoint is zero
+				// hash and max index.
+				PreviousOutPoint: *wire.NewOutPoint(&chainhash.Hash{},
+					wire.MaxPrevOutIndex, wire.TxTreeRegular),
+				Sequence:        wire.MaxTxInSequenceNum,
+				BlockHeight:     wire.NullBlockHeight,
+				BlockIndex:      wire.NullBlockIndex,
+				SignatureScript: tspendScript,
+			})
+
+			return msgTx
+		},
+		is:       IsTSpend,
+		expected: true,
+		check:    checkTSpend,
+	}}
+
+	for i, test := range tests {
+		if got := test.is(test.createTx()); got != test.expected {
+			// Obtain error
+			err := test.check(test.createTx())
+			t.Fatalf("%v %v: failed got %v want %v error %v",
+				i, test.name, got, test.expected, err)
+		}
+	}
+}
+
+// tspendTxInNoPubkey
+var tspendTxInNoPubkey = wire.TxIn{
+	PreviousOutPoint: wire.OutPoint{
+		Hash:  chainhash.Hash{},
+		Index: 0xffffffff,
+		Tree:  wire.TxTreeRegular,
+	},
+	SignatureScript: []byte{
+		0xc2, // OP_TSPEND
+	},
+	BlockHeight: wire.NullBlockHeight,
+	BlockIndex:  wire.NullBlockIndex,
+	Sequence:    0xffffffff,
+}
+
+// tspendTxInInvalidPubkey is a TxIn with an invalid key on the OP_TSPEND.
+var tspendTxInInvalidPubkey = wire.TxIn{
+	PreviousOutPoint: wire.OutPoint{
+		Hash:  chainhash.Hash{},
+		Index: 0xffffffff,
+		Tree:  wire.TxTreeRegular,
+	},
+	SignatureScript: []byte{
+		0xc2, // OP_TSPEND
+		0x23, // OP_DATA_35
+		0x03, // Valid pubkey version
+		0x00, // invalid compressed key
+	},
```

Comment:
I am going to comment only on this one so that I don't repeat myself but this applies to all "test has more than one thing wrong with it" type comments you made.

Consensus and signature detection is inherently ordered. With that assumption in mind, I wrote the tests to be valid up to the point I wanted to hit. While this is somewhat confusing, as you pointed out, it does have the benefit that tests do bomb if you don't hit the expected error.

I am ok revisiting these tests at a later time if this is not acceptable to you but for now I don't want to spend days on this while we are trying to get this out the door.